include/ignition/math/detail/AxisIndex.hh

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef IGNITION_MATH_DETAIL_AXIS_INDEX_LOOKUP_FIELD_HH_
+#define IGNITION_MATH_DETAIL_AXIS_INDEX_LOOKUP_FIELD_HH_
+
+#include <algorithm>
+#include <map>
+#include <optional>
+#include <vector>
+
+#include <iostream>
+
+namespace ignition
+{
+  namespace math
+  {
+    // Inline bracket to help doxygen filtering.
+    inline namespace IGNITION_MATH_VERSION_NAMESPACE {
+
+      /// \brief Represents a sparse number line which can be searched via
+      /// search for indices.
+      template <typename T>
+      class AxisIndex
+      {
+        /// \brief Binary tree with indices
+        private: std::map<T, std::size_t> axisIndex;
+
+        /// \brief Number of indices
+        private: int numIndices{0};
+
+        /// \brief Register the existance of a measurement at _value.
+        /// \param[in] _value The position of the measurement.
+        public: void AddIndexIfNotFound(T _value)
+        {
+          if (axisIndex.find(_value) == axisIndex.end())
+          {
+            axisIndex[_value] = numIndices;
+            numIndices++;
+          }
+        }
+
+        /// \brief Get the number of unique indices.
+        /// \return The number of unique indices.
+        public: std::size_t GetNumUniqueIndices()
+        {
+          return axisIndex.size();
+        }
+
+        /// \brief Get the index of a measurement
+        /// \param[in] _value The position of the measurement.
+        /// \return The index of the measurement if found else return nullopt.
+        public: std::optional<std::size_t> GetIndex(T _value) const
+        {
+          auto res = axisIndex.find(_value);
+          if (res == axisIndex.end())
+          {
+            return std::nullopt;
+          }
+          else
+          {
+            return res->second;
+          }
+        }
+
+        /// \brief Get interpolators for a measurement.
+        /// \param[in] _value The position of the measurement.
+        /// \param[in] _tol The tolerance for the search.
+        /// \return The indices of the measurements that should be used for
+        /// interpolation. If the value is out of range, an empty vector is
+        /// returned. If the value is exact, a vector with a single index is
+        /// returned otherwise return the two indices which should be used for
+        /// interpolation.
+        public: std::vector<std::size_t> GetInterpolators(T _value,
+          double _tol = 1e-6) const
+        {
+          // Performs a BST to find the first element that is greater than or
+          // equal to the value.
+          auto it = axisIndex.lower_bound(_value);
+          if (it == axisIndex.end())
+          {
+            // Out of range
+            return {};
+          }
+          else if (fabs(it->first - _value) < _tol)
+          {
+            // Exact match
```

Comment:
@arjo129 meta: why introduce a tolerance to check for almost exact matches when we can return all points and defer to the interpolation algorithm to judge?

Reply:
So if you see how the `GetInterpolators` works it returns a variable number of points. By doing this this way it makes for cleaner interpolation code as we can separate the kind of interpolation we want to use (none, linear, bilinear, trilinear) just based on the length of the return vector. If we were to go with the other alternative we would spend a lot more time debugging the corner cases of the interpolation.

Reply:
@arjo129 hmm, why special casing? Isn't linear (resp. bilinear) interpolation just trilinear interpolation for a point that is on an edge (resp. face) of the 3D volumes that make up its grid? I guess I'm missing some numerical nuances here. Anywho, so long as the tolerance can be adjusted I'm OK with this.

Reply:
See #412 for my original plan.